nemo/lightning/pytorch/callbacks/model_checkpoint.py

```diff
@@ -450,10 +484,42 @@ def _save_checkpoint(self, trainer: 'pytorch_lightning.Trainer', filepath: str)
                 TrainerContext.from_trainer(trainer).io_dump(ckpt_to_dir(filepath) / "context")
 
             if self.async_save:
+                self._last_checkpoint_saved = filepath
                 logging.info(f'Scheduled async checkpoint save for {filepath}')
             else:
                 finalize_fn()
 
+    def _save_last_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[str, torch.Tensor]) -> None:
```

Comment:
I'm wondering if there is some way to avoid overriding the whole method. This is always risky, since we lose touch with the upstream.

How is our flow different from the one in PTL which makes us add `saved_current_step` logic and also not rely on `self.last_model_path`?
Is it because PTL links to _any_ available last checkpoint (not necessarily from the last iteration)?

Reply:
Yeah, I made two changes:
1. Made sure to add a symlink only when the current step was actually saved. As you suggested, PTL always links to the last checkpoint saved, which might not correspond to the latest step
2. Added [these lines](https://github.com/NVIDIA/NeMo/blob/200cae3a3794f539db80186c1c5fc40ca9723995/nemo/lightning/pytorch/callbacks/model_checkpoint.py#L505-L508) which fix `last_model_path` saved to the `*-last` checkpoint state dict when using symlinks

I'll think about whether we can make these fixes without overwriting the entire `_save_last_checkpoint` method

Reply:
Thanks. Maybe overwriting save_last_checkpoint is inevitable in which case current version is ok

Reply:
Great, thanks, this is great.

Do you know how `last_model_path` is used during restart? I'm wondering if it will be valid if e.g. failure happens between the regular and "last" ckpt save

Reply:
I think `last_model_path` is only used when removing the previous `-last` model to ensure we only retain a single `-last` checkpoint. If failure happens between the regular and last checkpoint save, I **don't** think the state dict will be valid, but I also don't think this is a concern, because we'd end up restoring from the previously saved `-last` checkpoint which does have the correct state dict.